In [1]:
sc

SparkContext 

 Spark UI 

 
 Version 
 v2.4.3 
 Master 
 local[8] 
 AppName 
 Databricks Shell

# Load Marketing Qualified Leads Data

In [3]:
# File location and type
file_location = "/FileStore/tables/olist_marketing_qualified_leads_dataset.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
mql = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(mql)

mql_id first_contact_date landing_page_id origin dac32acd4db4c29c230538b72f8dd87d 2018-02-01 88740e65d5d6b056e0cda098e1ea6313 social 8c18d1de7f67e60dbd64e3c07d7e9d5d 2017-10-20 007f9098284a86ee80ddeb25d53e0af8 paid_search b4bc852d233dfefc5131f593b538befa 2018-03-22 a7982125ff7aa3b2054c6e44f9d28522 organic_search 6be030b81c75970747525b843c1ef4f8 2018-01-22 d45d558f0daeecf3cccdffe3c59684aa email 5420aad7fec3549a85876ba1c529bd84 2018-02-21 b48ec5f3b04e9068441002a19df93c6c organic_search 28bdfd5f057764b54c38770f95c69f2f 2018-01-14 22c29808c4f815213303f8933030604c organic_search 126a0d10becbaafcb2e72ce6848cf32c 2018-05-15 6a110e795dd487f1cf8d7583671987af email f76136f54d14a3345951f25b7932366b 2018-05-24 d51b0d02f063ba1d053db6d97226eec3 email 2f838cade4a6012a6cb1016d1d8d95ed 2017-11-10 aeac92c0f5ae22a04ed3b746cce3a1b6 organic_search 7281942387a1a0c3f72a50a8b0bb0920 2017-12-25 88740e65d5d6b056e0cda098e1ea6313 social c534e2f89539e866011e1e75adfefbfb 2018-01-24 f017be4dbf86243af5c1ebed0cff36a2 unknown c3764ab3fed8fbf556c87fdeaea744a9 2018-04-30 35c9b150ab36fe584c1f24fd458c453a paid_search a555fb36b9368110ede0f043dfc3b9a0 2018-04-04 2f08471dfb83baef522f952438a36f6e referral 988c9372560f30a5f28f92640bb2b3da 2018-05-23 73f31a40697cc90a86c1273563ac230e organic_search 327174d3648a2d047e8940d7d15204ca 2018-04-03 1ceb590cd1e00c7ee95220971f82693d organic_search a5448d7164426cc7f3260a7b2001e3e6 2017-12-08 bdaeb752e5c3caa215ebe04c61e22325 paid_search 6dda8f32d9ddc2a9ffe87be7d361a5c9 2018-02-16 b76ef37428e6799c421989521c0e5077 unknown be8f254ef25c5a40ac8b395ed4fc6033 2018-01-13 1d4ddc25ccaa2adeded340b6eda66182 organic_search 3183fc21acb07e30098c42484e12d697 2018-01-15 22c29808c4f815213303f8933030604c direct_traffic 8051b43253152a39fe9367c69b79b42b 2018-02-23 a7982125ff7aa3b2054c6e44f9d28522 organic_search 443a4c031f05c42f729a1ff2ac219b5a 2017-07-25 3cd2a8301d5395f21cdd3094c9eaa2f4 paid_search e1b036022a6aaa7d187caae163216533 2017-07-13 caef563a428392ddc7807d2a529413b9 organic_search 29860f5735fc4987b5f8ef3ee2767847 2018-03-12 a7982125ff7aa3b2054c6e44f9d28522 unknown 4bdc2dfb783f8c24c6d0d032a623febc 2018-04-14 22c29808c4f815213303f8933030604c organic_search 6d1ca411fcbc6a21484877d59be651db 2017-12-31 22c29808c4f815213303f8933030604c display b12b646e693cbf411f8eaae5204ffdfe 2017-09-11 b48ec5f3b04e9068441002a19df93c6c organic_search e3154fc901362537cc34b26fd18f0f05 2018-05-29 ece960c5dab1a015cc1cb5a419b0c63b referral f101b87d3b9adb48f2016bf059067ea2 2018-05-02 d83b0d0e48c8447d1d5507a44027a955 organic_search a8b32485ea3dafd7c06f87afe4f7c6d2 2018-04-15 22c29808c4f815213303f8933030604c social 2f0bf07b716733d56ffa0b68cfe3823f 2018-05-08 30534107aa0013db92899c96a1017d0b social 676c0c9be43fd984c92b0ecfdab3c0d9 2017-12-31 15cf89bab2c3163b4f9d8601d108f8b2 display 5f62b80e297edbcbfb1bb211062834e9 2018-01-26 22c29808c4f815213303f8933030604c paid_search b62db8fadd1b177cc5ed1d239d6e2f1e 2018-02-22 a7982125ff7aa3b2054c6e44f9d28522 organic_search f0b8024455e6c0cb3ab7495325afee5c 2018-05-11 58326e62183c14b0c03085c33b9fdc44 social 2f20c4826f9fabdb206537363642dd65 2018-04-18 0218f6be0b76aca72ab4d00ee9e8cf10 paid_search 65855f77bb6ee720c0a940b9cf8101a9 2017-09-26 e492ee5eaf1697716985cc6f33f9cd9b referral 91ae48805e8e96055d454f043054059b 2018-03-07 f017be4dbf86243af5c1ebed0cff36a2 display 899c617012fd3225c396958f91b21f84 2018-03-16 b76ef37428e6799c421989521c0e5077 unknown 365aa59e779547d7e081b220f8fa67ac 2017-09-29 f017be4dbf86243af5c1ebed0cff36a2 organic_search f5fee8f7da74f4887f5bcae2bafb6dd6 2018-01-14 22c29808c4f815213303f8933030604c paid_search 82ade9609d9915ca5e830df04b7d90e2 2018-01-03 22c29808c4f815213303f8933030604c referral 1d833a78ef4b542f594ef2cd734e1d14 2018-05-30 d823616cbcb84fff516db22f366caaa5 paid_search 9cfc99e322c9ce9f0fcfe1843113b472 2018-03-28 ea0650174e273c41e12002dcdd4708be direct_traffic 2b6319cb1c507418504e28bf285072ee 2018-02-05 f017be4dbf86243af5c1ebed0cff36a2 organic_search 4c721b2017d5fb477659ba48277446de 2018-02-17 5209e2e679d32a6619feb69

In [4]:
print(mql.count(), "rows") # Row Count
print(len(mql.columns), "Columns") # Column Count

8000 rows
4 Columns

In [5]:
# Create a view or table

#temp_table_name = "olist_marketing_qualified_leads_dataset_csv"

#df.createOrReplaceTempView(temp_table_name)

In [6]:
%sql

/* Query the created temp table in a SQL cell */

--select * from `olist_marketing_qualified_leads_dataset_csv`

In [7]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.

#permanent_table_name = "olist_marketing_qualified_leads_dataset_csv"

# df.write.format("parquet").saveAsTable(permanent_table_name)

# Load Closed Deals Dataset

In [9]:
# File location and type
file_location = "/FileStore/tables/olist_closed_deals_dataset.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
cd = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(cd)

mql_id seller_id sdr_id sr_id won_date business_segment lead_type lead_behaviour_profile has_company has_gtin average_stock business_type declared_product_catalog_size declared_monthly_revenue 5420aad7fec3549a85876ba1c529bd84 2c43fb513632d29b3b58df74816f1b06 a8387c01a09e99ce014107505b92388c 4ef15afb4b2723d8f3d81e51ec7afefe 2018-02-26 19:58:54 pet online_medium cat null null null reseller null 0.0 a555fb36b9368110ede0f043dfc3b9a0 bbb7d7893a450660432ea6652310ebb7 09285259593c61296eef10c734121d5b d3d1e91a157ea7f90548eef82f1955e3 2018-05-08 20:17:59 car_accessories industry eagle null null null reseller null 0.0 327174d3648a2d047e8940d7d15204ca 612170e34b97004b3ba37eae81836b4c b90f87164b5f8c2cfa5c8572834dbe3f 6565aa9ce3178a5caf6171827af3a9ba 2018-06-05 17:27:23 home_appliances online_big cat null null null reseller null 0.0 f5fee8f7da74f4887f5bcae2bafb6dd6 21e1781e36faf92725dde4730a88ca0f 56bf83c4bb35763a51c2baab501b4c67 d3d1e91a157ea7f90548eef82f1955e3 2018-01-17 13:51:03 food_drink online_small null null null null reseller null 0.0 ffe640179b554e295c167a2f6be528e0 ed8cb7b190ceb6067227478e48cf8dde 4b339f9567d060bcea4f5136b9f5949e d3d1e91a157ea7f90548eef82f1955e3 2018-07-03 20:17:45 home_appliances industry wolf null null null manufacturer null 0.0 b94fba7670eeb44dce2a0d8eb790e9f5 1c742ac33582852aaf3bcfbf5893abcf fdb16d3cbbeb5798f2f66c4096be026d 495d4e95a8cf8bbf8b432b612a2aa328 2018-02-07 18:04:05 health_beauty online_medium null null null null manufacturer null 0.0 c3e30ed7ac989117c7e1e719b4ac128f 92d7568ad0c5c76fd7d341b2d46f24d6 4b339f9567d060bcea4f5136b9f5949e 85fc447d336637ba1df43e793199fbc8 2018-04-16 18:18:22 computers online_medium wolf null null null reseller null 0.0 b02c89251106e1fdd9d92744be9f94f2 44ed138eca6214d572ce1d813fb0049b 34d40cdaf94010a1d05b0d6212f9e909 4ef15afb4b2723d8f3d81e51ec7afefe 2018-04-17 17:01:57 health_beauty offline cat null null null manufacturer null 0.0 a90a37898cc5f2718385a2fb981caaff 0b28859cd04d23edefee9c591fb03cd8 f42a2bd194f7802ab052a815c8de65b7 6565aa9ce3178a5caf6171827af3a9ba 2018-05-14 18:37:15 household_utilities offline eagle null null null reseller null 0.0 0173e8d8b1d94a355b440fb67388f532 87d73636a3acf123e842bb890a4db036 9d12ef1a7eca3ec58c545c678af7869c 9e4d1098a3b0f5da39b0bc48f9876645 2018-04-24 03:00:00 household_utilities online_medium eagle null null null manufacturer null 0.0 4e4e70d504b4c0006c8287dedc99d0fc f7a0d94e966c5665355a182d5b199fcf fdb16d3cbbeb5798f2f66c4096be026d 4ef15afb4b2723d8f3d81e51ec7afefe 2018-02-28 19:42:14 construction_tools_house_garden online_medium cat null null null reseller null 0.0 408a9c4a79800232ac656249af3162eb b566ab0ef88016e00422755e305103c6 de63de0d10a6012430098db33c679b0b d3d1e91a157ea7f90548eef82f1955e3 2018-02-23 13:54:40 construction_tools_house_garden online_big cat null null null reseller null 0.0 c556184b3fe2087834850b68fa435cee 2d2322d842118867781fc737e96d59a1 09285259593c61296eef10c734121d5b 2695de1affa7750089c0455f8ce27021 2018-05-22 13:24:08 household_utilities online_medium cat null null null reseller null 0.0 67455a0673f5cf26deae5b270c003e83 e7012030d0fdd1d3ca504f6de7909c35 4b339f9567d060bcea4f5136b9f5949e 9ae085775a198122c5586fa830ff7f2b 2018-06-14 14:30:20 toys online_big wolf null null null reseller null 0.0 ab11f58a6bf7036a656364b0ba68fb74 9e7c5f4d7770eab65738cca38f9efccf 068066e24f0c643eb1d089c7dd20cd73 de63de0d10a6012430098db33c679b0b 2018-04-10 03:00:00 sports_leisure online_medium cat null null null reseller null 0.0 bae38661a27f6228ba38c36e766ed769 6a6b1614baaaf766293c17d8cb8c5a9d f42a2bd194f7802ab052a815c8de65b7 495d4e95a8cf8bbf8b432b612a2aa328 2018-09-10 14:18:01 stationery industry cat null null 20-50 reseller 2000.0 100000.0 450b97a48568c3118ff397f73abe4e4a df91910b6a03bb2e3358fa6a35e32f6f 09285259593c61296eef10c734121d5b 060c0a26f19f4d66b42e0d8796688490 2018-04-10 19:21:54 food_drink online_medium cat null null null manufacturer null 0.0 9878253904482726e741460a99adba86 92d46311e4fa7583d14c351fdc881af6 de63de0d10a6012430098

In [10]:
print(cd.count(), "rows") # Row Count
print(len(cd.columns), "Columns") # Column Count

842 rows
14 Columns

In [11]:
# Create a view or table

#temp_table_name = "olist_closed_deals_dataset_csv"

#df.createOrReplaceTempView(temp_table_name)

In [12]:
%sql

/* Query the created temp table in a SQL cell */

--select * from `olist_closed_deals_dataset_csv`

In [13]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.

#permanent_table_name = "olist_closed_deals_dataset_csv"

# df.write.format("parquet").saveAsTable(permanent_table_name)

# Sales Performance Overview
After a MQL filled a form on landing page to sign up for seller, a Sales Development Representative(SDR) contacted the MQL and gathered more information about the lead. Then a Sales Representative(SR) consulted the MQL. So interaction between SDRs/SRs and MQLs can affect conversion from MQLs to sellers.

At this section we will deal with two aspects of sales result, conversion rate and sales length.

Note: A MQL who finally signed up for seller is called a _closed deal_.

In [15]:
# Merge 'mql' with 'cd'
# Merge by 'left' in order to evaluate conversion rate

mql_cd = mql.join(cd, mql.mql_id == cd.mql_id,how='left').drop(cd.mql_id)

display(mql_cd)

mql_id first_contact_date landing_page_id origin seller_id sdr_id sr_id won_date business_segment lead_type lead_behaviour_profile has_company has_gtin average_stock business_type declared_product_catalog_size declared_monthly_revenue dac32acd4db4c29c230538b72f8dd87d 2018-02-01 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null 8c18d1de7f67e60dbd64e3c07d7e9d5d 2017-10-20 007f9098284a86ee80ddeb25d53e0af8 paid_search null null null null null null null null null null null null null b4bc852d233dfefc5131f593b538befa 2018-03-22 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null 6be030b81c75970747525b843c1ef4f8 2018-01-22 d45d558f0daeecf3cccdffe3c59684aa email null null null null null null null null null null null null null 5420aad7fec3549a85876ba1c529bd84 2018-02-21 b48ec5f3b04e9068441002a19df93c6c organic_search 2c43fb513632d29b3b58df74816f1b06 a8387c01a09e99ce014107505b92388c 4ef15afb4b2723d8f3d81e51ec7afefe 2018-02-26 19:58:54 pet online_medium cat null null null reseller null 0.0 28bdfd5f057764b54c38770f95c69f2f 2018-01-14 22c29808c4f815213303f8933030604c organic_search null null null null null null null null null null null null null 126a0d10becbaafcb2e72ce6848cf32c 2018-05-15 6a110e795dd487f1cf8d7583671987af email null null null null null null null null null null null null null f76136f54d14a3345951f25b7932366b 2018-05-24 d51b0d02f063ba1d053db6d97226eec3 email null null null null null null null null null null null null null 2f838cade4a6012a6cb1016d1d8d95ed 2017-11-10 aeac92c0f5ae22a04ed3b746cce3a1b6 organic_search null null null null null null null null null null null null null 7281942387a1a0c3f72a50a8b0bb0920 2017-12-25 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null c534e2f89539e866011e1e75adfefbfb 2018-01-24 f017be4dbf86243af5c1ebed0cff36a2 unknown null null null null null null null null null null null null null c3764ab3fed8fbf556c87fdeaea744a9 2018-04-30 35c9b150ab36fe584c1f24fd458c453a paid_search null null null null null null null null null null null null null a555fb36b9368110ede0f043dfc3b9a0 2018-04-04 2f08471dfb83baef522f952438a36f6e referral bbb7d7893a450660432ea6652310ebb7 09285259593c61296eef10c734121d5b d3d1e91a157ea7f90548eef82f1955e3 2018-05-08 20:17:59 car_accessories industry eagle null null null reseller null 0.0 988c9372560f30a5f28f92640bb2b3da 2018-05-23 73f31a40697cc90a86c1273563ac230e organic_search null null null null null null null null null null null null null 327174d3648a2d047e8940d7d15204ca 2018-04-03 1ceb590cd1e00c7ee95220971f82693d organic_search 612170e34b97004b3ba37eae81836b4c b90f87164b5f8c2cfa5c8572834dbe3f 6565aa9ce3178a5caf6171827af3a9ba 2018-06-05 17:27:23 home_appliances online_big cat null null null reseller null 0.0 a5448d7164426cc7f3260a7b2001e3e6 2017-12-08 bdaeb752e5c3caa215ebe04c61e22325 paid_search null null null null null null null null null null null null null 6dda8f32d9ddc2a9ffe87be7d361a5c9 2018-02-16 b76ef37428e6799c421989521c0e5077 unknown null null null null null null null null null null null null null be8f254ef25c5a40ac8b395ed4fc6033 2018-01-13 1d4ddc25ccaa2adeded340b6eda66182 organic_search null null null null null null null null null null null null null 3183fc21acb07e30098c42484e12d697 2018-01-15 22c29808c4f815213303f8933030604c direct_traffic null null null null null null null null null null null null null 8051b43253152a39fe9367c69b79b42b 2018-02-23 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null 443a4c031f05c42f729a1ff2ac219b5a 2017-07-25 3cd2a8301d5395f21cdd3094c9eaa2f4 paid_search null null null null null null null null null null null null null e1b036022a6aaa7d187caae163216533 2017-07-13 caef563a428392ddc7807d2a529413b9 organic_search null null null null null null null null null null null null null 29860f5735fc4987b5f8ef3ee2767847 2018-03-

In [16]:
print(mql_cd.count(), "rows") # Row Count
print(len(mql_cd.columns), "Columns") # Column Count

8000 rows
17 Columns

In [17]:
# Add a column to distinguish signed MOLs from MQLs who left without signing up
mql_cd_1 = mql_cd.withColumn("seller_id_bool", mql_cd.seller_id.isNotNull())
display(mql_cd_1)

mql_id first_contact_date landing_page_id origin seller_id sdr_id sr_id won_date business_segment lead_type lead_behaviour_profile has_company has_gtin average_stock business_type declared_product_catalog_size declared_monthly_revenue seller_id_bool dac32acd4db4c29c230538b72f8dd87d 2018-02-01 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null false 8c18d1de7f67e60dbd64e3c07d7e9d5d 2017-10-20 007f9098284a86ee80ddeb25d53e0af8 paid_search null null null null null null null null null null null null null false b4bc852d233dfefc5131f593b538befa 2018-03-22 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null false 6be030b81c75970747525b843c1ef4f8 2018-01-22 d45d558f0daeecf3cccdffe3c59684aa email null null null null null null null null null null null null null false 5420aad7fec3549a85876ba1c529bd84 2018-02-21 b48ec5f3b04e9068441002a19df93c6c organic_search 2c43fb513632d29b3b58df74816f1b06 a8387c01a09e99ce014107505b92388c 4ef15afb4b2723d8f3d81e51ec7afefe 2018-02-26 19:58:54 pet online_medium cat null null null reseller null 0.0 true 28bdfd5f057764b54c38770f95c69f2f 2018-01-14 22c29808c4f815213303f8933030604c organic_search null null null null null null null null null null null null null false 126a0d10becbaafcb2e72ce6848cf32c 2018-05-15 6a110e795dd487f1cf8d7583671987af email null null null null null null null null null null null null null false f76136f54d14a3345951f25b7932366b 2018-05-24 d51b0d02f063ba1d053db6d97226eec3 email null null null null null null null null null null null null null false 2f838cade4a6012a6cb1016d1d8d95ed 2017-11-10 aeac92c0f5ae22a04ed3b746cce3a1b6 organic_search null null null null null null null null null null null null null false 7281942387a1a0c3f72a50a8b0bb0920 2017-12-25 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null false c534e2f89539e866011e1e75adfefbfb 2018-01-24 f017be4dbf86243af5c1ebed0cff36a2 unknown null null null null null null null null null null null null null false c3764ab3fed8fbf556c87fdeaea744a9 2018-04-30 35c9b150ab36fe584c1f24fd458c453a paid_search null null null null null null null null null null null null null false a555fb36b9368110ede0f043dfc3b9a0 2018-04-04 2f08471dfb83baef522f952438a36f6e referral bbb7d7893a450660432ea6652310ebb7 09285259593c61296eef10c734121d5b d3d1e91a157ea7f90548eef82f1955e3 2018-05-08 20:17:59 car_accessories industry eagle null null null reseller null 0.0 true 988c9372560f30a5f28f92640bb2b3da 2018-05-23 73f31a40697cc90a86c1273563ac230e organic_search null null null null null null null null null null null null null false 327174d3648a2d047e8940d7d15204ca 2018-04-03 1ceb590cd1e00c7ee95220971f82693d organic_search 612170e34b97004b3ba37eae81836b4c b90f87164b5f8c2cfa5c8572834dbe3f 6565aa9ce3178a5caf6171827af3a9ba 2018-06-05 17:27:23 home_appliances online_big cat null null null reseller null 0.0 true a5448d7164426cc7f3260a7b2001e3e6 2017-12-08 bdaeb752e5c3caa215ebe04c61e22325 paid_search null null null null null null null null null null null null null false 6dda8f32d9ddc2a9ffe87be7d361a5c9 2018-02-16 b76ef37428e6799c421989521c0e5077 unknown null null null null null null null null null null null null null false be8f254ef25c5a40ac8b395ed4fc6033 2018-01-13 1d4ddc25ccaa2adeded340b6eda66182 organic_search null null null null null null null null null null null null null false 3183fc21acb07e30098c42484e12d697 2018-01-15 22c29808c4f815213303f8933030604c direct_traffic null null null null null null null null null null null null null false 8051b43253152a39fe9367c69b79b42b 2018-02-23 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null false 443a4c031f05c42f729a1ff2ac219b5a 2017-07-25 3cd2a8301d5395f21cdd3094c9eaa2f4 paid_search null null null null null null null null null null null null null false e1b036022a6aaa7d187caae163216533 2017-07-13 caef563a428392dd

In [18]:
# Add a 'year-month' column
import re
from pyspark.sql.functions import udf
@udf
def date_Y_M(date):
  #print(text)
  return re.search('\d{4}-\d{2}', str(date)).group()

mql_cd_2 = mql_cd_1.withColumn("first_contact_date_Y_M", date_Y_M(mql_cd_1.first_contact_date))
display(mql_cd_2)

mql_id first_contact_date landing_page_id origin seller_id sdr_id sr_id won_date business_segment lead_type lead_behaviour_profile has_company has_gtin average_stock business_type declared_product_catalog_size declared_monthly_revenue seller_id_bool first_contact_date_Y_M dac32acd4db4c29c230538b72f8dd87d 2018-02-01 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null false 2018-02 8c18d1de7f67e60dbd64e3c07d7e9d5d 2017-10-20 007f9098284a86ee80ddeb25d53e0af8 paid_search null null null null null null null null null null null null null false 2017-10 b4bc852d233dfefc5131f593b538befa 2018-03-22 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null false 2018-03 6be030b81c75970747525b843c1ef4f8 2018-01-22 d45d558f0daeecf3cccdffe3c59684aa email null null null null null null null null null null null null null false 2018-01 5420aad7fec3549a85876ba1c529bd84 2018-02-21 b48ec5f3b04e9068441002a19df93c6c organic_search 2c43fb513632d29b3b58df74816f1b06 a8387c01a09e99ce014107505b92388c 4ef15afb4b2723d8f3d81e51ec7afefe 2018-02-26 19:58:54 pet online_medium cat null null null reseller null 0.0 true 2018-02 28bdfd5f057764b54c38770f95c69f2f 2018-01-14 22c29808c4f815213303f8933030604c organic_search null null null null null null null null null null null null null false 2018-01 126a0d10becbaafcb2e72ce6848cf32c 2018-05-15 6a110e795dd487f1cf8d7583671987af email null null null null null null null null null null null null null false 2018-05 f76136f54d14a3345951f25b7932366b 2018-05-24 d51b0d02f063ba1d053db6d97226eec3 email null null null null null null null null null null null null null false 2018-05 2f838cade4a6012a6cb1016d1d8d95ed 2017-11-10 aeac92c0f5ae22a04ed3b746cce3a1b6 organic_search null null null null null null null null null null null null null false 2017-11 7281942387a1a0c3f72a50a8b0bb0920 2017-12-25 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null false 2017-12 c534e2f89539e866011e1e75adfefbfb 2018-01-24 f017be4dbf86243af5c1ebed0cff36a2 unknown null null null null null null null null null null null null null false 2018-01 c3764ab3fed8fbf556c87fdeaea744a9 2018-04-30 35c9b150ab36fe584c1f24fd458c453a paid_search null null null null null null null null null null null null null false 2018-04 a555fb36b9368110ede0f043dfc3b9a0 2018-04-04 2f08471dfb83baef522f952438a36f6e referral bbb7d7893a450660432ea6652310ebb7 09285259593c61296eef10c734121d5b d3d1e91a157ea7f90548eef82f1955e3 2018-05-08 20:17:59 car_accessories industry eagle null null null reseller null 0.0 true 2018-04 988c9372560f30a5f28f92640bb2b3da 2018-05-23 73f31a40697cc90a86c1273563ac230e organic_search null null null null null null null null null null null null null false 2018-05 327174d3648a2d047e8940d7d15204ca 2018-04-03 1ceb590cd1e00c7ee95220971f82693d organic_search 612170e34b97004b3ba37eae81836b4c b90f87164b5f8c2cfa5c8572834dbe3f 6565aa9ce3178a5caf6171827af3a9ba 2018-06-05 17:27:23 home_appliances online_big cat null null null reseller null 0.0 true 2018-04 a5448d7164426cc7f3260a7b2001e3e6 2017-12-08 bdaeb752e5c3caa215ebe04c61e22325 paid_search null null null null null null null null null null null null null false 2017-12 6dda8f32d9ddc2a9ffe87be7d361a5c9 2018-02-16 b76ef37428e6799c421989521c0e5077 unknown null null null null null null null null null null null null null false 2018-02 be8f254ef25c5a40ac8b395ed4fc6033 2018-01-13 1d4ddc25ccaa2adeded340b6eda66182 organic_search null null null null null null null null null null null null null false 2018-01 3183fc21acb07e30098c42484e12d697 2018-01-15 22c29808c4f815213303f8933030604c direct_traffic null null null null null null null null null null null null null false 2018-01 8051b43253152a39fe9367c69b79b42b 2018-02-23 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null false 2018-02 443a4c031f05c42f729a1ff2ac219b5a 2017

In [19]:
# Another way to Add a 'year-month' column
#from pyspark.sql.functions import expr
#mql_cd_2 = mql_cd_1.withColumn("first_contact_date_Y_M", expr("to_date(first_contact_date, 'yyyy-MM')"))
#display(mql_cd_2)

In [20]:
# Using sqlContext.sql or spark.sql to write SQL instead of sql.functions

mql_cd_2.createOrReplaceTempView("mql_cd_2")

monthly_cd = spark.sql("select first_contact_date_Y_M, count(case when seller_id_bool=True then 1 end) as number_of_closed_deals from mql_cd_2 \
group by first_contact_date_Y_M \
order by first_contact_date_Y_M")
display(monthly_cd)

first_contact_date_Y_M number_of_closed_deals 2017-06 0 2017-07 2 2017-08 9 2017-09 7 2017-10 14 2017-11 18 2017-12 11 2018-01 152 2018-02 149 2018-03 167 2018-04 183 2018-05 130

#### Monthly volume of closed deals sharply increased in 2018.

In [22]:
# Another way to Compute monthly closed deals using sql.functions
#from pyspark.sql.functions import col,when,count
#monthly_cd = mql_cd_2.groupby("first_contact_date_Y_M")\
#.agg(count(when(col("seller_id_bool"), True)).alias("number_of_closed_deals"))\
#.orderBy("first_contact_date_Y_M")
#display(monthly_cd)

## Conversion Rate
Conversion rate means the percentage of MQLs who finally signed up for sellers (closed deals).

In [24]:
# Calculate monthly conversion rate
monthly_conversion = spark.sql("select first_contact_date_Y_M, \
count(case when seller_id_bool=True then 1 end) as number_of_closed_deals, \
count(first_contact_date_Y_M) as number_of_mqls, \
round(count(case when seller_id_bool=True then 1 end)/count(first_contact_date_Y_M)*100, 1) as conversion_rate from mql_cd_2 \
group by first_contact_date_Y_M \
order by first_contact_date_Y_M")
display(monthly_conversion)

first_contact_date_Y_M number_of_closed_deals number_of_mqls conversion_rate 2017-06 0 4 0.0 2017-07 2 239 0.8 2017-08 9 386 2.3 2017-09 7 312 2.2 2017-10 14 416 3.4 2017-11 18 445 4.0 2017-12 11 200 5.5 2018-01 152 1141 13.3 2018-02 149 1028 14.5 2018-03 167 1174 14.2 2018-04 183 1352 13.5 2018-05 130 1303 10.0

#### Monthly conversion rate increased with volume.

## Sales Length
Sales length means period from first contact to signing up for seller.

In [27]:
from pyspark.sql.functions import datediff, to_date

mql_cd_3 = mql_cd_2.withColumn("sales_length",datediff(to_date(mql_cd_2.won_date, 'yyyy-MM-dd'),to_date(mql_cd_2.first_contact_date, 'yyyy-MM-dd')))
mql_cd_3.createOrReplaceTempView("mql_cd_3")
display(mql_cd_3)

mql_id first_contact_date landing_page_id origin seller_id sdr_id sr_id won_date business_segment lead_type lead_behaviour_profile has_company has_gtin average_stock business_type declared_product_catalog_size declared_monthly_revenue seller_id_bool first_contact_date_Y_M sales_length dac32acd4db4c29c230538b72f8dd87d 2018-02-01 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null false 2018-02 null 8c18d1de7f67e60dbd64e3c07d7e9d5d 2017-10-20 007f9098284a86ee80ddeb25d53e0af8 paid_search null null null null null null null null null null null null null false 2017-10 null b4bc852d233dfefc5131f593b538befa 2018-03-22 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null null null null null null null null null null null false 2018-03 null 6be030b81c75970747525b843c1ef4f8 2018-01-22 d45d558f0daeecf3cccdffe3c59684aa email null null null null null null null null null null null null null false 2018-01 null 5420aad7fec3549a85876ba1c529bd84 2018-02-21 b48ec5f3b04e9068441002a19df93c6c organic_search 2c43fb513632d29b3b58df74816f1b06 a8387c01a09e99ce014107505b92388c 4ef15afb4b2723d8f3d81e51ec7afefe 2018-02-26 19:58:54 pet online_medium cat null null null reseller null 0.0 true 2018-02 5 28bdfd5f057764b54c38770f95c69f2f 2018-01-14 22c29808c4f815213303f8933030604c organic_search null null null null null null null null null null null null null false 2018-01 null 126a0d10becbaafcb2e72ce6848cf32c 2018-05-15 6a110e795dd487f1cf8d7583671987af email null null null null null null null null null null null null null false 2018-05 null f76136f54d14a3345951f25b7932366b 2018-05-24 d51b0d02f063ba1d053db6d97226eec3 email null null null null null null null null null null null null null false 2018-05 null 2f838cade4a6012a6cb1016d1d8d95ed 2017-11-10 aeac92c0f5ae22a04ed3b746cce3a1b6 organic_search null null null null null null null null null null null null null false 2017-11 null 7281942387a1a0c3f72a50a8b0bb0920 2017-12-25 88740e65d5d6b056e0cda098e1ea6313 social null null null null null null null null null null null null null false 2017-12 null c534e2f89539e866011e1e75adfefbfb 2018-01-24 f017be4dbf86243af5c1ebed0cff36a2 unknown null null null null null null null null null null null null null false 2018-01 null c3764ab3fed8fbf556c87fdeaea744a9 2018-04-30 35c9b150ab36fe584c1f24fd458c453a paid_search null null null null null null null null null null null null null false 2018-04 null a555fb36b9368110ede0f043dfc3b9a0 2018-04-04 2f08471dfb83baef522f952438a36f6e referral bbb7d7893a450660432ea6652310ebb7 09285259593c61296eef10c734121d5b d3d1e91a157ea7f90548eef82f1955e3 2018-05-08 20:17:59 car_accessories industry eagle null null null reseller null 0.0 true 2018-04 34 988c9372560f30a5f28f92640bb2b3da 2018-05-23 73f31a40697cc90a86c1273563ac230e organic_search null null null null null null null null null null null null null false 2018-05 null 327174d3648a2d047e8940d7d15204ca 2018-04-03 1ceb590cd1e00c7ee95220971f82693d organic_search 612170e34b97004b3ba37eae81836b4c b90f87164b5f8c2cfa5c8572834dbe3f 6565aa9ce3178a5caf6171827af3a9ba 2018-06-05 17:27:23 home_appliances online_big cat null null null reseller null 0.0 true 2018-04 63 a5448d7164426cc7f3260a7b2001e3e6 2017-12-08 bdaeb752e5c3caa215ebe04c61e22325 paid_search null null null null null null null null null null null null null false 2017-12 null 6dda8f32d9ddc2a9ffe87be7d361a5c9 2018-02-16 b76ef37428e6799c421989521c0e5077 unknown null null null null null null null null null null null null null false 2018-02 null be8f254ef25c5a40ac8b395ed4fc6033 2018-01-13 1d4ddc25ccaa2adeded340b6eda66182 organic_search null null null null null null null null null null null null null false 2018-01 null 3183fc21acb07e30098c42484e12d697 2018-01-15 22c29808c4f815213303f8933030604c direct_traffic null null null null null null null null null null null null null false 2018-01 null 8051b43253152a39fe9367c69b79b42b 2018-02-23 a7982125ff7aa3b2054c6e44f9d28522 organic_search null null null 

In [28]:
# Count closed deals by sales length for year 2017
sales_length_2017 = spark.sql("select sales_length, \
count(case when seller_id_bool=True then 1 end) as number_of_closed_deals from mql_cd_3 \
where first_contact_date_Y_M like '2017%' \
group by sales_length \
order by sales_length")
sales_length_2017.createOrReplaceTempView("sales_length_2017")
display(sales_length_2017)

sales_length number_of_closed_deals null 0 4 1 14 1 20 1 32 1 40 1 41 1 43 1 44 1 54 1 59 1 69 1 77 1 84 1 90 1 107 1 127 1 130 1 141 1 151 1 161 1 162 1 165 1 171 1 174 1 179 1 183 1 188 1 190 1 194 1 206 1 216 1 223 1 226 1 237 1 247 1 254 1 267 2 296 1 306 3 312 1 316 1 317 1 321 2 325 1 328 1 329 1 335 1 336 1 340 1 341 3 353 1 369 1 371 1 373 1 427 1

In [29]:
# Count closed deals by sales length for year 2018
sales_length_2018 = spark.sql("select sales_length, \
count(case when seller_id_bool=True then 1 end) as number_of_closed_deals from mql_cd_3 \
where first_contact_date_Y_M like '2018%' \
group by sales_length \
order by sales_length")
sales_length_2018.createOrReplaceTempView("sales_length_2018")
display(sales_length_2018)

sales_length number_of_closed_deals null 0 -2 1 0 33 1 43 2 35 3 29 4 32 5 43 6 32 7 35 8 23 9 23 10 25 11 22 12 15 13 20 14 15 15 13 16 11 17 16 18 8 19 9 20 7 21 9 22 10 23 11 24 6 25 4 26 3 27 2 28 5 29 4 30 8 31 3 32 4 33 3 34 4 35 8 36 3 37 5 38 5 39 1 40 4 41 3 42 5 43 2 44 1 45 2 46 2 47 3 48 3 49 3 50 3 51 1 53 1 54 1 55 3 56 4 57 2 58 3 60 3 62 1 63 3 65 1 66 2 68 1 69 2 70 1 72 3 73 2 74 2 75 2 76 2 77 2 79 2 80 2 81 2 84 3 85 2 86 1 87 3 88 2 91 1 92 2 93 1 94 1 96 1 97 3 98 1 99 1 100 2 101 2 102 1 103 1 106 2 108 1 109 1 110 2 111 2 112 1 113 3 114 1 116 1 117 1 119 1 120 1 124 1 125 1 126 4 127 1 130 1 131 1 132 2 134 1 137 1 142 1 144 2 145 1 148 3 153 1 155 2 156 2 162 1 163 3 164 1 166 1 169 1 172 1 174 1 175 2 178 1 180 1 181 1 182 1 184 1 187 1 189 1 191 1 192 1 194 1 195 1 198 1 199 1 200 1 202 1 203 1 204 1 206 1 207 1 210 3 211 1 216 1 228 1 236 1 241 2 252 1 260 2 292 1 306 1

In [30]:
# Calculate running total of closed deals by sales length for year 2017
# NOT efficient replace it later
sales_length_2017_running_total = spark.sql("select a.sales_length, \
a.number_of_closed_deals, \
sum(b.number_of_closed_deals) from sales_length_2017 as a \
join sales_length_2017 as b \
on (a.sales_length >= b.sales_length) \
group by a.sales_length, a.number_of_closed_deals \
order by a.sales_length")
display(sales_length_2017_running_total)

sales_length number_of_closed_deals sum(number_of_closed_deals) 4 1 1 14 1 2 20 1 3 32 1 4 40 1 5 41 1 6 43 1 7 44 1 8 54 1 9 59 1 10 69 1 11 77 1 12 84 1 13 90 1 14 107 1 15 127 1 16 130 1 17 141 1 18 151 1 19 161 1 20 162 1 21 165 1 22 171 1 23 174 1 24 179 1 25 183 1 26 188 1 27 190 1 28 194 1 29 206 1 30 216 1 31 223 1 32 226 1 33 237 1 34 247 1 35 254 1 36 267 2 38 296 1 39 306 3 42 312 1 43 316 1 44 317 1 45 321 2 47 325 1 48 328 1 49 329 1 50 335 1 51 336 1 52 340 1 53 341 3 56 353 1 57 369 1 58 371 1 59 373 1 60 427 1 61

In [31]:
# Calculate running total of closed deals by sales length for year 2018
# NOT efficient replace it later
sales_length_2018_running_total = spark.sql("select a.sales_length, \
a.number_of_closed_deals, \
sum(b.number_of_closed_deals) from sales_length_2018 as a \
join sales_length_2018 as b \
on (a.sales_length >= b.sales_length) \
group by a.sales_length, a.number_of_closed_deals \
order by a.sales_length")
display(sales_length_2018_running_total)

sales_length number_of_closed_deals sum(number_of_closed_deals) -2 1 1 0 33 34 1 43 77 2 35 112 3 29 141 4 32 173 5 43 216 6 32 248 7 35 283 8 23 306 9 23 329 10 25 354 11 22 376 12 15 391 13 20 411 14 15 426 15 13 439 16 11 450 17 16 466 18 8 474 19 9 483 20 7 490 21 9 499 22 10 509 23 11 520 24 6 526 25 4 530 26 3 533 27 2 535 28 5 540 29 4 544 30 8 552 31 3 555 32 4 559 33 3 562 34 4 566 35 8 574 36 3 577 37 5 582 38 5 587 39 1 588 40 4 592 41 3 595 42 5 600 43 2 602 44 1 603 45 2 605 46 2 607 47 3 610 48 3 613 49 3 616 50 3 619 51 1 620 53 1 621 54 1 622 55 3 625 56 4 629 57 2 631 58 3 634 60 3 637 62 1 638 63 3 641 65 1 642 66 2 644 68 1 645 69 2 647 70 1 648 72 3 651 73 2 653 74 2 655 75 2 657 76 2 659 77 2 661 79 2 663 80 2 665 81 2 667 84 3 670 85 2 672 86 1 673 87 3 676 88 2 678 91 1 679 92 2 681 93 1 682 94 1 683 96 1 684 97 3 687 98 1 688 99 1 689 100 2 691 101 2 693 102 1 694 103 1 695 106 2 697 108 1 698 109 1 699 110 2 701 111 2 703 112 1 704 113 3 707 114 1 708 116 1 709 117 1 710 119 1 711 120 1 712 124 1 713 125 1 714 126 4 718 127 1 719 130 1 720 131 1 721 132 2 723 134 1 724 137 1 725 142 1 726 144 2 728 145 1 729 148 3 732 153 1 733 155 2 735 156 2 737 162 1 738 163 3 741 164 1 742 166 1 743 169 1 744 172 1 745 174 1 746 175 2 748 178 1 749 180 1 750 181 1 751 182 1 752 184 1 753 187 1 754 189 1 755 191 1 756 192 1 757 194 1 758 195 1 759 198 1 760 199 1 761 200 1 762 202 1 763 203 1 764 204 1 765 206 1 766 207 1 767 210 3 770 211 1 771 216 1 772 228 1 773 236 1 774 241 2 776 252 1 777 260 2 779 292 1 780 306 1 781

##### Sales length was dramatically shortened in 2018 as well.
While few deals 8 out of 61 (around 13%) were closed within 50 days in 2017, 616 out of 781 (around 79%) were closed within 50 days in 2018.
##### Generally, both conversion rate and sales length were improved in 2018 against 2017.